File: src/types/report.ts
export interface ProjectReportRow {
  id: string
  name: string
  status: string
  progress: number
  budget: number
  spent: number
  percentUsed: number
  milestonesTotal: number
  milestonesCompleted: number
  punchOpen: number
}

export interface PortfolioReport {
  projectsTotal: number
  projectsByStatus: Record<string, number>
  totalBudget: number
  totalSpent: number
  budgetPercentUsed: number
  milestonesTotal: number
  milestonesCompleted: number
  milestonesOverdue: number
  onTimeRate: number | null
  punchOpen: number
  punchResolved: number
  projects: ProjectReportRow[]
}

export type ProjectCsvReport = 'milestones' | 'cost-entries' | 'punch' | 'site-logs'

export const PROJECT_CSV_REPORTS: { value: ProjectCsvReport; label: string }[] = [
  { value: 'milestones', label: 'Milestones' },
  { value: 'cost-entries', label: 'Cost entries' },
  { value: 'punch', label: 'Punch list' },
  { value: 'site-logs', label: 'Site logs' },
]

File: backend/app/schemas/report.py
import uuid
from decimal import Decimal
from pydantic import BaseModel


class ProjectReportRow(BaseModel):
    id: uuid.UUID
    name: str
    status: str
    progress: int
    budget: Decimal
    spent: Decimal
    percent_used: float
    milestones_total: int
    milestones_completed: int
    punch_open: int


class PortfolioReport(BaseModel):
    projects_total: int
    projects_by_status: dict[str, int]
    total_budget: Decimal
    total_spent: Decimal
    budget_percent_used: float
    milestones_total: int
    milestones_completed: int
    milestones_overdue: int
    on_time_rate: float | None  # % of completed milestones done by their due date; None if no data
    punch_open: int
    punch_resolved: int
    projects: list[ProjectReportRow]
